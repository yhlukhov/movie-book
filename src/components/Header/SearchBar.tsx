import { useState, FormEvent } from 'react'
import { Form, SearchInput, SearchButton, Title } from './styled'

export const SearchBar = () => {
  const [input, setInput] = useState('')

  const handleSearch = (e:FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    console.log(input)
  }

  return (
    <>
      <Title>Find your movie</Title>
      <Form onSubmit={handleSearch}>
        <SearchInput
          type='text'
          value={input}
          placeholder='What do you want to watch?'
          onChange={(e) => setInput(e.target.value)}
        />
        <SearchButton>Search</SearchButton>
      </Form>
    </>
  )
}
