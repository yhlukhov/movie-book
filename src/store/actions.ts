export {setMovie} from './movieSlice'
export {setSearchTerm, setGenre, setSortBy} from './appSlice'