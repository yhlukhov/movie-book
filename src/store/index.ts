export { default } from './store'

export * from './store'
export * from './appSlice'
export * from './movieSlice'