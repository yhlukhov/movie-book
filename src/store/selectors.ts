export {selectSearchTerm, selectGenre, selectSortBy} from './appSlice'
export {selectMovies, selectMovie} from './movieSlice'