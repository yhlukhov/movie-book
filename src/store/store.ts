import {combineReducers, configureStore} from '@reduxjs/toolkit'
import appReducer from './appSlice'
import movieReducer from './movieSlice'

const rootReducer = combineReducers({
  app: appReducer,
  movies: movieReducer,
})

export type RootState = ReturnType<typeof rootReducer>

export default configureStore({
  reducer: rootReducer
})