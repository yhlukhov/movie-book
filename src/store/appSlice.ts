import {createSlice, PayloadAction} from '@reduxjs/toolkit'
import { GenreType, SortTypes } from '../types'
import { RootState } from './store'

const initialState = {
  sortBy: 'release_date' as SortTypes,
  genre:'all' as GenreType|'all',
  searchInput:'',
  searchTerm: '',
}

const appSlice = createSlice({
  name: 'app',
  initialState,
  reducers: {
    setSearchInput:(state, action:PayloadAction<string>) => {
      state.searchInput = action.payload
    },
    setSearchTerm: (state, action:PayloadAction<string>) => {
      state.searchTerm = action.payload
    },
    setGenre: (state, action:PayloadAction<GenreType|'all'>) => {
      state.genre = action.payload
    },
    setSortBy: (state, action:PayloadAction<SortTypes>) => {
      state.sortBy = action.payload
    }
  }
})

// Action creators:
export const {setSearchInput, setSearchTerm, setGenre, setSortBy} = appSlice.actions

// Selectors:
export const selectSearchInput = (store:RootState) => store.app.searchInput
export const selectSearchTerm = (store:RootState) => store.app.searchTerm
export const selectGenre = (store:RootState) => store.app.genre
export const selectSortBy = (store:RootState) => store.app.sortBy

export default appSlice.reducer