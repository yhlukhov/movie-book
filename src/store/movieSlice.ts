import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { MovieFormData, MovieType } from '../types'
import { DispatchType, RootState, GetState } from './store'
import { loadMovies, loadMoviesBySearchTerm, postMovie, putMovie, deleteMovie, loadMovie } from '../API'
import { acquireSortOrder } from '../utilities'

const initialState = {
  movies: [] as Array<MovieType>,
  movie: null as null|MovieType,
  movieHints: [] as Array<MovieType>,
  addMovieTitle:'',
  totalAmount: 0,
  limit: 12,
  page: 1,
}

// Movie Reducer Slice
export const movieSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setMovie: (state, action:PayloadAction<MovieType|null>) => {
      state.movie = action.payload
    },
    setMovies: (state, action:PayloadAction<Array<MovieType>>) => {
      state.movies = action.payload
    },
    setMovieHints: (state, action:PayloadAction<Array<MovieType>>) => {
      state.movieHints = action.payload
    },
    setTotalAmount: (state, action:PayloadAction<number>) => {
      state.totalAmount = action.payload
    },
    setPage: (state, action:PayloadAction<number>) => {
      state.page = action.payload
    },
    setAddMovieTitle: (state, action:PayloadAction<string>) => {
      state.addMovieTitle = action.payload
    }
  }
})

// Action Creators:
export const {setMovie, setMovies, setMovieHints, setTotalAmount,setPage, setAddMovieTitle} = movieSlice.actions

// Selectors:
export const selectMovie = (state:RootState) => state.movies.movie
export const selectMovies = (state:RootState) => state.movies.movies
export const selectMovieHints = (state:RootState) => state.movies.movieHints
export const selectTotalAmount = (state:RootState) => state.movies.totalAmount
export const selectLimit = (state:RootState) => state.movies.limit
export const selectPage = (state: RootState) => state.movies.page
export const selectAddMovieTitle = (state:RootState) => state.movies.addMovieTitle

// Thunks:
export const getMovies = () => async (dispatch:DispatchType, getState:GetState) => {
  
  const state = getState()
  const {genre, sortBy} = state.app
  const {page, limit} = state.movies
  const sortOrder = acquireSortOrder(sortBy)
  const genreVal = genre === 'all' ? null : genre
  const response = await loadMovies(genreVal, sortBy, sortOrder, page, limit)
  if (response) {
    dispatch(setMovies(response.data))
    dispatch(setTotalAmount(response.totalAmount))
  }
}

export const searchMovies = (searchTerm:string) => async (dispatch:DispatchType, getState:GetState) => {
  const state = getState()
  const {genre, sortBy} = state.app
  const {page, limit} = state.movies
  const sortOrder = acquireSortOrder(sortBy)
  const genreVal = genre === 'all' ? null : genre
  const response = await loadMoviesBySearchTerm(searchTerm, genreVal, sortBy, sortOrder, page, limit)
  if (response) {
    dispatch(setMovies(response.data)) 
    dispatch(setTotalAmount(response.totalAmount))
  }
}

export const getMovie = (id:string) => async (dispatch:DispatchType, getState:GetState) => {
    const response = await loadMovie(id)
    if(response) {
      dispatch(setMovie(response))
    }
}

export const getMovieHints = (searchTerm:string) => async (dispatch: DispatchType, getState:GetState) => {
  const state = getState()
  const {genre, sortBy} = state.app
  const sortOrder = acquireSortOrder(sortBy)
  const genreVal = genre === 'all' ? null : genre
  const response = await loadMoviesBySearchTerm(searchTerm, genreVal, sortBy,sortOrder, 1, 6)
  if (response) {
    dispatch(setMovieHints(response.data))
  }
}

export const addMovie = (data:MovieFormData) => async (dispatch:DispatchType, getState:GetState) => {
  await postMovie(data)
  dispatch(setAddMovieTitle(data.title))
  const searchTerm = getState().app.searchTerm
  if (searchTerm) {
    dispatch(searchMovies(searchTerm))
    return
  }
  dispatch(getMovies())
}

export const editMovie = (movie:MovieType) => async (dispatch:DispatchType, getState:GetState) => {
  if(!movie.tagline) movie.tagline = movie.title
  await putMovie(movie)
  const searchTerm = getState().app.searchTerm
  if (searchTerm) {
    dispatch(searchMovies(searchTerm))
    return
  }
  dispatch(getMovies())
}

export const removeMovie = (id:string) => async(dispatch:DispatchType, getState:GetState) => {
  await deleteMovie(id)
  const searchTerm = getState().app.searchTerm
  if (searchTerm) {
    dispatch(searchMovies(searchTerm))
    return
  }
  dispatch(getMovies())
}

export default movieSlice.reducer