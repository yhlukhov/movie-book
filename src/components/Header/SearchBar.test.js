import { render, screen, fireEvent } from '@testing-library/react';
import { SearchBar } from './SearchBar';
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer';
import store from '../../store'
const WithProviderRouter = (component) => <Provider store={store}><Router>{component}</Router></Provider>

describe('Search Bar', () => {
  beforeEach(() => {
    render(WithProviderRouter(<SearchBar />))
  })

  test('Rendres correctly',  () => {
    const tree = renderer.create(WithProviderRouter(<SearchBar />)).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Renders Search Bar elements', () => {
    expect(screen.getByPlaceholderText(/What do you want to watch/i)).toBeInTheDocument()
    expect(screen.getByRole('textbox')).toBeInTheDocument()
  })
  
  test('Assures Search input is empty', () => {
    expect(screen.getByRole('textbox')).toBeEmptyDOMElement()
  })

  test('Write text into Search input', () => {
    fireEvent.change(screen.getByRole('textbox'), {
      target: {value: 'Call Me by Your Name'}
    })
    expect(screen.getByRole('textbox')).toHaveValue('Call Me by Your Name')
  })

  test('Search button is rendered', () => {
    expect(screen.getByRole('button', {name: 'Search'})).toBeInTheDocument() 
  })

  test('Click on Search btn submits the form', () => {
    const onSubmit = jest.fn()
    const form = screen.getByTestId('search-form')
    const searchBtn = screen.getByRole('button', {name: 'Search'})
    form.onsubmit = onSubmit
    expect(form).toBeInTheDocument()
    fireEvent.click(searchBtn) 
    expect(onSubmit).toBeCalled()
  })
})