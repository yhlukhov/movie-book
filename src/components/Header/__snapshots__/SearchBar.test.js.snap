// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar Rendres correctly 1`] = `
Array [
  <div
    className="sc-jSgvzq kopskl"
  >
    Find your movie
  </div>,
  <form
    className="sc-gKseQn iDZLLA"
    data-testid="search-form"
    onSubmit={[Function]}
  >
    <label>
      <input
        className="sc-iBPTik fQCWZd"
        onChange={[Function]}
        placeholder="What do you want to watch?"
        type="text"
        value=""
      />
      
    </label>
    <button
      className="sc-fubCzh fZjCWK"
    >
      Search
    </button>
  </form>,
]
`;
